std-bits/base/src/main/java/org/enso/base/parser/FormatDetectingNumberParser.java

```diff
@@ -0,0 +1,301 @@
+  }
+
+  public record NumberParseLong(long number, String symbol) implements NumberParseResultSuccess {
+    @Override
+    public NumberParseResultSuccess negate() {
+      return new NumberParseLong(-number, symbol);
+    }
+
+    @Override
+    public NumberParseResultSuccess withSymbol(String symbol) {
+      return new NumberParseLong(number, symbol);
+    }
+  }
+
+  public record NumberParseDouble(double number, String symbol)
+      implements NumberParseResultSuccess {
+    @Override
+    public NumberParseResultSuccess negate() {
+      return new NumberParseDouble(-number, symbol);
+    }
+
+    @Override
+    public NumberParseResultSuccess withSymbol(String symbol) {
+      return new NumberParseDouble(number, symbol);
+    }
+  }
+
+  public record NumberParseFailure(String message) implements NumberParseResult {}
+
+  private final boolean allowSymbol;
+  private final boolean allowLeadingZeroes;
+  private final boolean allowLeadingTrailingWhitespace;
+  private NegativeSign negativeSign;
+  private NumberWithSeparators numberWithSeparators;
+
+  public FormatDetectingNumberParser() {
+    this(true, true, true, NegativeSign.UNKNOWN, NumberWithSeparators.UNKNOWN);
+  }
+
+  public FormatDetectingNumberParser(
+      boolean allowSymbol,
+      boolean allowLeadingZeroes,
+      boolean allowLeadingTrailingWhitespace,
+      NegativeSign negativeSign,
+      NumberWithSeparators numberWithSeparators) {
+    this.allowSymbol = allowSymbol;
+    this.allowLeadingZeroes = allowLeadingZeroes;
+    this.allowLeadingTrailingWhitespace = allowLeadingTrailingWhitespace;
+    this.negativeSign = negativeSign;
+    this.numberWithSeparators = numberWithSeparators;
+  }
+
+  public NegativeSign negativeSign() {
+    return negativeSign;
+  }
+
+  public void setNegativeSign(NegativeSign newNegativeSign) {
+    if (negativeSign != NegativeSign.UNKNOWN) {
+      throw new IllegalStateException("Negative Sign Already Set.");
+    }
+    negativeSign = newNegativeSign;
+  }
+
+  public NumberWithSeparators numberWithSeparators() {
+    return numberWithSeparators;
+  }
+
+  /**
+   * Parse a string into a number.
+   *
+   * @param value the string to parse.
+   * @param integer whether to parse a Long or a Double.
+   * @return the parsed number, or a failure if the parse was unsuccessful.
+   */
+  public NumberParseResult parse(CharSequence value, boolean integer) {
+    // State
+    boolean lastWasWhitespace = false;
+    boolean encounteredContent = false;
+    boolean encounteredSign = false;
+    boolean isNegative = false;
+    NumberParseResultSuccess number = null;
+    String symbol = "";
+
+    // Scan the value
+    int idx = 0;
+    int length = value.length();
+    while (idx < length) {
+      char c = value.charAt(idx);
+
+      if (Character.isWhitespace(c)) {
+        if (!allowLeadingTrailingWhitespace && !encounteredContent) {
+          return new NumberParseFailure("Leading Whitespace.");
```

Comment:
Perhaps
```suggestion
          return new NumberParseFailure("Unexpected leading whitespace.");
```
will be more user friendly?

As leading whitespace is not in itself an error, it's only a problem if we expect it to not be there.